collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java

```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
+
+            if(userSpinnerValues.contains(getResources().getString(R.string.all_forms))) {
+                userSpinnerValues.remove(getResources().getString(R.string.all_forms));
+                sortedSpinnerValues.add(getResources().getString(R.string.all_forms));
+            }
+
+            if(userSpinnerValues.contains(userFormsKey)) {
+                userSpinnerValues.remove(userFormsKey);
+                sortedSpinnerValues.add(userFormsKey);
+            }
+
+            sortedSpinnerValues.addAll(userSpinnerValues);
+
+            ArrayAdapter<String> userArrayAdapter = new ArrayAdapter<String>(this, android.R
+                    .layout.simple_spinner_item, sortedSpinnerValues);
+            userArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+            formAccountsSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
+                @Override
+                public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
+                    String selectedKey = parent.getItemAtPosition(position).toString();
+                    HashMap<String, HashMap<String, FormDetails>> userForms =
+                            splitFormsIntoAccounts(getBaseContext(), mFormNamesAndURLs);
+
+                    if(userForms.containsKey(selectedKey)) {
+                        refreshFormList(userForms.get(selectedKey));
                     }
-                    mFormList.add(j, item);
                 }
-            }
-            selectSupersededForms();
-            mFormListAdapter.notifyDataSetChanged();
+
+                @Override
+                public void onNothingSelected(AdapterView<?> parent) {}
+            });
+            formAccountsSpinner.setAdapter(userArrayAdapter);
+
             mDownloadButton.setEnabled(!(selectedItemCount() == 0));
         }
     }
 
+    private void refreshFormList(HashMap<String, FormDetails> forms) {
+        mFormList.clear();
+
+        ArrayList<String> ids = new ArrayList<String>(forms.keySet());
+        for (int i = 0; i < forms.size(); i++) {
+            String formDetailsKey = ids.get(i);
+            FormDetails details = forms.get(formDetailsKey);
+            HashMap<String, String> item = new HashMap<String, String>();
+            item.put(FORMNAME, details.formName);
+            item.put(FORMID_DISPLAY,
+                    ((details.formVersion == null) ? "" : (getString(R.string.version) + " " + details.formVersion + " ")) +
+                            "ID: " + details.formID );
+            item.put(FORMDETAIL_KEY, formDetailsKey);
+            item.put(FORM_ID_KEY, details.formID);
+            item.put(FORM_VERSION_KEY, details.formVersion);
+
+            // Insert the new form in alphabetical order.
+            if (mFormList.size() == 0) {
+                mFormList.add(item);
+            } else {
+                int j;
+                for (j = 0; j < mFormList.size(); j++) {
```

Comment:
if you insert into the list then sort this happens in O(n + nlogn) = O(n(1 + logn)) = O(nlogn) (upper bound w/mergesort for example), I think your code here will be O(n * n / 2) = O(n^2/2) = O(n^2), so that's significantly worse.

tl;dr, remove all this and sort the list after they've all been inserted